### PR TITLE
[microNPU][3] Plan generation for the cascader

### PR DESCRIPTION
**Pull request by mbaret**
RFC: apache/tvm-rfcs#37
Issue: #9429

The cascader creates 'Plans' which describe how to schedule subgraphs. As part of the cascading algorithm, it's necessary to explore a large variety of Plans which are Pareto optimal (in terms of memory usage and performance). This is done by the Plan generation algorithm.

This commit adds the TensorConfig and Plan data structures which hold information on how to schedule the tensors/operators. Additionally, it includes functions to calculate Pareto frontiers which are used to cull sub-optimal Plans.

### CODE REVIEW

File: python/tvm/contrib/ethosu/cascader/pareto.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+"""Pareto optimisation functions for the NPU cascader."""
+from typing import List
+
+from tvm import Object
+
+from . import _ffi_api
+from .plan import Plan
+
+
+def get_pareto_frontier(costs: List[List[float]]) -> List[bool]:
```
Comment (manupa-arm) [inline python/tvm/contrib/ethosu/cascader/pareto.py:26]: Should we add docstrings for these functions ?

File: python/tvm/contrib/ethosu/cascader/pareto.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    for i, cost in enumerate(costs):
+        for j, value in enumerate(cost):
+            costs[i][j] = float(value)
+
+    return [bool(v) for v in _ffi_api.GetParetoFrontier(costs)]
+
+
+def thin_vector(vec: List[Object], max_size: int) -> List[Object]:
```
Comment (manupa-arm) [inline python/tvm/contrib/ethosu/cascader/pareto.py:34]: Should we add docstrings for these functions ?

File: python/tvm/contrib/ethosu/cascader/pareto.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    return list(_ffi_api.ThinVector(vec, max_size))
+
+
+def pareto_cull_plans(plans: List[Plan], max_plans: int) -> List[Plan]:
```
Comment (manupa-arm) [inline python/tvm/contrib/ethosu/cascader/pareto.py:38]: Should we add docstrings for these functions ?

File: python/tvm/contrib/ethosu/cascader/plan.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+from tvm.runtime import Object
+
+from . import _ffi_api
+from .graph import Tensor, Part
+from .tensor_config import TensorConfig, MemoryRegion
+
+
+@tvm._ffi.register_object("contrib.ethosu.cascader.Plan")
+class Plan(Object):
+    """Plan class"""
```
Comment (manupa-arm) [inline python/tvm/contrib/ethosu/cascader/plan.py:30]: We would need a better description than this.

File: python/tvm/contrib/ethosu/cascader/plan_generator.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+from tvm.contrib.ethosu.cascader.tensor_config import MemoryRegion
+
+from . import _ffi_api
+from .cascader_options import CascaderOptions
+from .plan import Plan
+from .stripe_config import StripeConfig
+from .graph import CascaderGraph, Part, Tensor
+
+
+def generate_output_stripe_configs(part: Part, stripe_factors: int) -> List[StripeConfig]:
```
Comment (manupa-arm) [inline python/tvm/contrib/ethosu/cascader/plan_generator.py:29]: Docstring here as well and for later functions ?

**Review by manupa-arm — COMMENTED** [review]
Broadly looks good!.

I noticed that we dont have docstrings for some python functions/classes/objects that calls into C++ version that is well documented. However, it feels better to write a short summary there and with a forward pointer to detailed description.

**Comment by mbaret** [comment]
I've added docs to all the functions/classes that are meant to be used/access via Python, and marked those functions which are used only for testing as private.

**Review by manupa-arm — APPROVED** [review]
Thanks @mbaret . LGTM!

**Comment by manupa-arm** [comment]
Thanks @mbaret . This is merged now!

**Pull request merged**